kotlinx-coroutines-core/common/src/channels/AbstractChannel.kt

```diff
@@ -381,27 +356,30 @@ internal abstract class AbstractSendChannel<E> : SendChannel<E> {
         while (true) {
             if (select.isSelected) return
             if (full) {
-                val enqueueOp = TryEnqueueSendDesc(element, select, block)
-                val enqueueResult = select.performAtomicIfNotSelected(enqueueOp) ?: return
-                when {
-                    enqueueResult === ALREADY_SELECTED -> return
-                    enqueueResult === ENQUEUE_FAILED -> {} // retry
-                    enqueueResult is Closed<*> -> throw recoverStackTrace(enqueueResult.sendException)
-                    else -> error("performAtomicIfNotSelected(TryEnqueueSendDesc) returned $enqueueResult")
-                }
-            } else {
-                val offerResult = offerSelectInternal(element, select)
-                when {
-                    offerResult === ALREADY_SELECTED -> return
-                    offerResult === OFFER_FAILED -> {} // retry
-                    offerResult === OFFER_SUCCESS -> {
-                        block.startCoroutineUnintercepted(receiver = this, completion = select.completion)
+                val node = SendSelect(element, this, select, block)
+                when (val enqueueResult = enqueueSend(node)) {
+                    null -> { // enqueued successfully
+                        select.disposeOnSelect(node)
                         return
                     }
-                    offerResult is Closed<*> -> throw recoverStackTrace(offerResult.sendException)
-                    else -> error("offerSelectInternal returned $offerResult")
+                    is Closed<*> -> {
+                        helpClose(enqueueResult)
```

Comment:
Similar `helpClose` is missing from `registerSelectSend`, making close non-linearizable with `onSend`

Reply:
`sendSuspend` and `registerSelectSend` look very similar, though it's not clear whether we can get rid of copy-paste here.
But this bug becomes obvious if these two methods are grouped together.

Ditto for receive

Reply:
Fixed `helpClose` in `registerSelectSend`. Postone sharing code to later optimization.